File: backend/app/services/sessions.py
import asyncio
import uuid
from typing import Dict, List, TypedDict

from app.core.config import SESSION_MAX_CHARS, SESSION_MAX_MESSAGES


class SessionMessage(TypedDict):
    role: str
    content: str


session_lock = asyncio.Lock()
sessions: Dict[str, List[SessionMessage]] = {}


async def create_session() -> str:
    session_id = f"session_{uuid.uuid4().hex}"
    async with session_lock:
        sessions[session_id] = []
    return session_id


async def get_session_messages(session_id: str) -> List[SessionMessage]:
    async with session_lock:
        return list(sessions.get(session_id, []))


async def append_session_message(
    session_id: str, role: str, content: str
) -> None:
    if not content.strip():
        return
    async with session_lock:
        messages = sessions.setdefault(session_id, [])
        messages.append({"role": role, "content": content})
        while len(messages) > SESSION_MAX_MESSAGES:
            messages.pop(0)
        total_chars = sum(len(msg["content"]) for msg in messages)
        while total_chars > SESSION_MAX_CHARS and messages:
            removed = messages.pop(0)
            total_chars -= len(removed["content"])


def format_session_prompt(history: List[SessionMessage], prompt: str) -> str:
    transcript: List[str] = []
    for message in history:
        role_label = "User" if message["role"] == "user" else "Assistant"
        transcript.append(f"{role_label}: {message['content']}")
    transcript.append(f"User: {prompt}")
    return "\n\n".join(transcript)

File: backend-bun/src/services/sessions.ts
import { config } from "../lib/config.ts";
import { generateId } from "../lib/time.ts";

interface SessionMessage {
  role: "user" | "assistant";
  content: string;
}

const sessions = new Map<string, SessionMessage[]>();

export function createSession(): string {
  const sessionId = generateId("session");
  sessions.set(sessionId, []);
  return sessionId;
}

export function getSessionMessages(sessionId: string): SessionMessage[] {
  return [...(sessions.get(sessionId) ?? [])];
}

export function appendSessionMessage(sessionId: string, role: "user" | "assistant", content: string): void {
  if (!content.trim()) return;

  let messages = sessions.get(sessionId);
  if (!messages) {
    messages = [];
    sessions.set(sessionId, messages);
  }

  messages.push({ role, content });

  // Enforce max messages limit
  while (messages.length > config.SESSION_MAX_MESSAGES) {
    messages.shift();
  }

  // Enforce max chars limit
  let totalChars = messages.reduce((sum, msg) => sum + msg.content.length, 0);
  while (totalChars > config.SESSION_MAX_CHARS && messages.length > 0) {
    const removed = messages.shift();
    if (removed) {
      totalChars -= removed.content.length;
    }
  }
}

export function formatSessionPrompt(history: SessionMessage[], prompt: string): string {
  const transcript: string[] = [];

  for (const message of history) {
    const roleLabel = message.role === "user" ? "User" : "Assistant";
    transcript.push(`${roleLabel}: ${message.content}`);
  }

  transcript.push(`User: ${prompt}`);
  return transcript.join("\n\n");
}
